specs/src/recovery.md

```diff
@@ -0,0 +1,525 @@
+
+While pull based gossip and broadcast trees are incredibly powerful tools for
+increasing throughput, it's easy to derive attacks that fail these constraints
+even with the assumptions. First we discuss those attacks, and then we will
+discuss strategies for mitigation.
+
+### Denial of Service
+
+The purpose of the strict diconnection rules are to prevent all scenarios where
+an attacker can flood the network with valid messages. If invalid messages are
+sent, then the attacker is kicked from their address book.
+
+Currently, there are no known gaps. Please comment if you find one.
+
+### Sybil Attacks
+
+#### Pull Based Gossip
+
+Upon requesting data in a standard pull based protocol, faulty peers can simply
+not send requested data. Typically, honest nodes add a timeout, after which they
+will request the data from a different peer. Given our assumption that nodes are
+not being eclipse attacked, nodes will eventually find an honest peer with the
+data.
+
+The problem arises when applications have less time to distribute data than they
+can avoid or remove an attacker. Recovery is such as case. If an attacker wants
+to disrupt a network, it only needs to slow the distribution of data
+sufficiently beyond the `ProposalTimeout`. While the network will eventually
+distribute that data, the throughput will be dramatically decreased until
+connections to malicious peers are removed.
+
+> Side note: its easy to confuse a congested peer as a malicious one. This makes
+> peer scoring incredibly difficult for pull based systems. If timeouts for
+> scoring or addional requests are too agressive, feedback loops emerge.
+> Congested nodes can hit timeouts, resulting in requesting more duplicate data,
+> which causes more congestion.
+
+**Conclusions:**
+
+- Time sensitive applications require additional mechanisms in order to use pull
+  based gossip.
+
+#### Broadcast Trees
+
+Using the most naive design of broadcast trees, where the broadcaster
+distributes equal amounts of block data to each of its peers, a single faulty
+peer will break the first constraint of recovery. Even if parity data is added,
+determined attackers can simply spin up a large majority of nodes in the network
+and not propagate data.
+
+**Conclusions:**
+
+1) The usage of broadcast trees requires mechanisms that allow for faulty nodes
+   to exist.
+2) Validators must be able to differentiate between themselves and
+   non-validators. Without this, attackers can also increase the number of nodes
+   in the network and poison the tree.
+
+### Defending Against Attacks
+
+In order to prevent attackers from sending `Have` messages to all peers and simply not responding, we use the `PerPeerConcurrentRequestCap` discussed above.
+
+```go
+// the largest number of validators to compose 1/3 of the voting power. If we
+// assume even voting power percentage, then that is simply 1/3 of the number of
+// validators.
+faultyValidators := max(1, validatorSet.Size() / 3)
```

Comment:
In practices this will still be stake weighted, right?

Also, would call this var sth like maxAllowedFaultyVals or sth like that.